File: Linked_List/430.Flatten_a_Multilevel_Doubly_Linked_List/Python/430.Flatten_a_Multilevel_Doubly_Linked_List.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution(object):
    def flatten(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#
        if (not head): #Issue/Boundary-case handling
            return None

        ############
        #Initialize
        ##### Result array #####
        res_arry = []

        ##### Previous head node #####
        prev_node = None


        ####################################
        #Recursion-based DFS loop traversal
        ##### Step 1: Recursion function call (i.e. head node) #####
        self.DFS_Traversal(head, res_arry)

        ##### Step 2: Post-process/Final completions #####
        len_res_arry = (len(res_arry) - 1) #Length of result array

        for res_arry_idx in range(len_res_arry):
            (res_arry[res_arry_idx].next) = res_arry[(res_arry_idx + 1)] #Keep updating/overwriting
            (res_arry[res_arry_idx].prev) = prev_node #Keep updating/overwriting
            (res_arry[res_arry_idx].child) = None #Keep updating/overwriting

            prev_node = res_arry[res_arry_idx] #Keep updating/overwriting

        res_arry[len_res_arry].next = None #Keep updating/overwriting
        res_arry[len_res_arry].prev = prev_node #Keep updating/overwriting
        res_arry[len_res_arry].child = None #Keep updating/overwriting

        return res_arry[0]
    

    def DFS_Traversal(self, head, res_arry):
        """
        :type head: Node
        :type res_arry: List[Node]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not head): #Issue/Boundary-case handling
            return

        res_arry.append(head) #Update/Record

        ####################
        #Whole process/flow
        self.DFS_Traversal((head.child), res_arry) #Recursion function call (i.e. child node)
        self.DFS_Traversal((head.next), res_arry) #Recursion function call (i.e. next node)

File: Linked_List/430.Flatten_a_Multilevel_Doubly_Linked_List/C++/430.Flatten_a_Multilevel_Doubly_Linked_List.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node * flatten(Node * head) 
    {
        //===============================//
        // Input type:                   //
        //   - Node * (head node)        //
        // Return type:                  //
        //   - Node * (result head node) //
        //===============================//
        if (! head) { return NULL; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Result vector /////
        vector<Node * > res_vect;

        ///// Length of result vector /////
        int len_res_vect;

        ///// Previous node /////
        Node * prev_node = NULL;


        /*Recursion-based DFS loop traversal*/
        ///// Step 1: Recursion function call (i.e. head node) /////
        dfs_traversal(head, res_vect);

        ///// Step 2: Post-process, final completions /////
        len_res_vect = res_vect.size(); //Length of result vector
        (len_res_vect--); //Update, Reduce

        for (int res_vect_idx = 0; (res_vect_idx < len_res_vect); (res_vect_idx++)) //Whole
        {
            ((res_vect[res_vect_idx])->next) = (res_vect[(res_vect_idx + 1)]); //Keep updating, overwriting
            ((res_vect[res_vect_idx])->prev) = prev_node; //Keep updating, overwriting
            ((res_vect[res_vect_idx])->child) = NULL; //Keep updating, overwriting

            prev_node = res_vect[res_vect_idx]; //Keep updating, overwriting

        } //Whole

        ((res_vect[len_res_vect])->next) = NULL; //Keep updating, overwriting
        ((res_vect[len_res_vect])->prev) = prev_node; //Keep updating, overwriting
        ((res_vect[len_res_vect])->child) = NULL; //Keep updating, overwriting

        return res_vect[0];
    }


    void dfs_traversal(Node * head, vector<Node * > & res_vect)
    {
        //==========================================//
        // Input type:                              //
        //   - Node * (head node)                   //
        //   - vector<Node * > ref. (result vector) //
        // Return type:                             //
        //   - void (no return)                     //
        //==========================================//
        if (! head) { return; } //Issue, Boundary-case handling

        res_vect.push_back(head); //Update, Record

        /*Whole process, flow*/
        dfs_traversal((head->child), res_vect); //Recursion function call (i.e. child node)
        dfs_traversal((head->next), res_vect); //Recursion function call (i.e. child node)
    }
};
